test/Quaternion.jl

```diff
@@ -267,6 +267,30 @@ Base.:(/)(a::MyReal, b::Real) = a.val / b
         @test isnan(Quaternion(1, 2, 3, NaN))
     end
 
+    @testset "*" begin
+        # verify basic correctness
+        q1 = Quaternion(1,0,0,0)
+        qi = Quaternion(0,1,0,0)
+        qj = Quaternion(0,0,1,0)
+        qk = Quaternion(0,0,0,1)
+        @test q1 * q1 == q1
+        @test q1 * qi == qi
+        @test q1 * qj == qj
+        @test q1 * qk == qk
+        @test qi * q1 == qi
+        @test qi * qi == -q1
+        @test qi * qj == qk
+        @test qi * qk == -qj
+        @test qj * q1 == qj
+        @test qj * qi == -qk
+        @test qj * qj == -q1
+        @test qj * qk == qi
+        @test qk * q1 == qk
+        @test qk * qi == qj
+        @test qk * qj == -qi
+        @test qk * qk == -q1
+    end
+
```

Comment:
```suggestion

    @testset "abs2" for _ in 1:100, T in (Float16, Float32, Float64)
        q = rand(Quaternion{T})
        @test abs2(q) == q'*q
    end

```
It would be nice to have these tests for

>The primary motivator was to ensure that q' * q == q * q' == abs2(q) exactly, even for float operations.

Reply:
I wasn't sure if we wanted to commit to this as a guarantee but I suppose there's no reason we wouldn't want it.